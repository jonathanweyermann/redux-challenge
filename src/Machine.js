import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Health from './health';
import './machine.css';
import { Row, Container, Col } from 'react-bootstrap';
import { updateMachine, updateMachineName } from '../src/actions';

function Machine (props) {
	const onNameChange = (event) => {
		props.updateMachineName(event.target.value);
	};

	const handleSubmit = (currentMachine) => {
		props.updateMachine({ ...currentMachine, name: props.machines.name}, props.machines.machines);
		event.preventDefault();
	};

	const machinesLayout = () => {
		let machine = props.machines.machines.find((machine) => machine.id === props.history.location.pathname.split('/')[2]);
		if (machine === undefined) {
			return null;
		}
		return (
			<Container className="container">
				<Row>
					<Col xs={6}>
						<h1>{machine.name}</h1>
						<h4>Update Device</h4>
						<form onSubmit={() => handleSubmit(machine)}>
							<label>
								Name:
								<input type="text" aria-label="name-input" value={props.machines.machineName} onChange={onNameChange} />
							</label>
							<input type="submit" value="Submit" />
						</form>
					</Col>
					<Col xs={6}>
						<div className="box"><Health value={machine.health} displayValue={true} /></div>
						<h2>Stats</h2>
						IP Address: {machine.ip_address}
					</Col>
				</Row>
			</Container>
		);
	};


	return (
		<React.Fragment>
			{ machinesLayout() }
		</React.Fragment>
	);
}

const mapStateToProps = function (state) {
	return {
		machines: state.machines,
		machineName: state.machineName
	};
};

Machine.propTypes = {
	machines: PropTypes.object,
	history: PropTypes.object,
	updateMachine: PropTypes.func,
	updateMachineName: PropTypes.func,
	machineName: PropTypes.string
};

export default connect(mapStateToProps, { updateMachine, updateMachineName })(Machine);
