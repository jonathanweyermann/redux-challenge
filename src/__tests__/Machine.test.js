
import { expect as chai_expect } from 'chai';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import React from 'react';
import { shallow, mount } from 'enzyme';
import Machine from '../Machine';
import { Provider } from 'react-redux';
import configureMockStore from "redux-mock-store";
import thunk from 'redux-thunk'
const props = {"history":{"length":5,"action":"POP","location":{"pathname":"/machines/99ade105-dee1-49eb-8ac4-e4d272f89fba","search":"","hash":"","key":"ybxgrt"}}};

const mockStore = configureMockStore([thunk]);
const store = mockStore(
  {"machines":{"machines":[{"id":"99ade105-dee1-49eb-8ac4-e4d272f89fba","name":"Renamed Machine","ip_address":"127.0.0.31","health":84},
													 {"id":"4111947a-6c58-4977-90fa-2caaaef88648","name":"Machine 2","ip_address":"127.0.0.4","health":95},
													 {"id":"57342663-909c-4adf-9829-6dd1a3aa9143","name":"Machine 3","ip_address":"127.0.0.55","health":2},
													 {"id":"5632e1ec-46cb-4895-bc8b-a91644568cd5","name":"Machine 4","ip_address":"127.0.0.3","health":22}],
							"machine":{"name":""},
							"name":""}}
);

describe('machine component', () => {
	const setUpFn = () => {
	 return mount(
		 <Provider store={store}>
			 <Machine {...props} />
		 </Provider>
	 );
 };

	let wrapper;
	beforeEach(() => {
		wrapper = setUpFn();
	});

	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});



const setup = () => {
  const utils = render(<Provider store={store}><Machine {...props} /></Provider>);
  const input = utils.getByLabelText('name-input')
  return {
    input,
    ...utils,
  }
}

test('renders name of current machine link', () => {
  const { getByText } = render(<Provider store={store}><Machine {...props} /></Provider>);
  const linkElement = getByText(/Renamed Machine/i);
  expect(linkElement).toBeInTheDocument();
});

test('renders name of new machine link once renamed', () => {
	const { input } = setup();
	fireEvent.change(input, { target: { value: 'Machine 12' } })
  const { getByText } = render(<Provider store={store}><Machine {...props} /></Provider>);

	setTimeout(function(){
		const linkElement = getByText(/Machine 12/i);
		expect(linkElement).toBeInTheDocument();
	}, 1000);

});
