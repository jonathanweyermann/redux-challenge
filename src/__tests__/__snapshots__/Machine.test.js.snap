// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`machine component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Machine)
    history={
      Object {
        "action": "POP",
        "length": 5,
        "location": Object {
          "hash": "",
          "key": "ybxgrt",
          "pathname": "/machines/99ade105-dee1-49eb-8ac4-e4d272f89fba",
          "search": "",
        },
      }
    }
  >
    <Machine
      history={
        Object {
          "action": "POP",
          "length": 5,
          "location": Object {
            "hash": "",
            "key": "ybxgrt",
            "pathname": "/machines/99ade105-dee1-49eb-8ac4-e4d272f89fba",
            "search": "",
          },
        }
      }
      machines={
        Object {
          "machine": Object {
            "name": "",
          },
          "machines": Array [
            Object {
              "health": 84,
              "id": "99ade105-dee1-49eb-8ac4-e4d272f89fba",
              "ip_address": "127.0.0.31",
              "name": "Renamed Machine",
            },
            Object {
              "health": 95,
              "id": "4111947a-6c58-4977-90fa-2caaaef88648",
              "ip_address": "127.0.0.4",
              "name": "Machine 2",
            },
            Object {
              "health": 2,
              "id": "57342663-909c-4adf-9829-6dd1a3aa9143",
              "ip_address": "127.0.0.55",
              "name": "Machine 3",
            },
            Object {
              "health": 22,
              "id": "5632e1ec-46cb-4895-bc8b-a91644568cd5",
              "ip_address": "127.0.0.3",
              "name": "Machine 4",
            },
          ],
          "name": "",
        }
      }
      updateMachine={[Function]}
      updateMachineName={[Function]}
    >
      <Container
        className="container"
        fluid={false}
      >
        <div
          className="container container"
        >
          <Row
            noGutters={false}
          >
            <div
              className="row"
            >
              <Col
                xs={6}
              >
                <div
                  className="col-6"
                >
                  <h1>
                    Renamed Machine
                  </h1>
                  <h4>
                    Update Device
                  </h4>
                  <form
                    onSubmit={[Function]}
                  >
                    <label>
                      Name:
                      <input
                        aria-label="name-input"
                        onChange={[Function]}
                        type="text"
                      />
                    </label>
                    <input
                      type="submit"
                      value="Submit"
                    />
                  </form>
                </div>
              </Col>
              <Col
                xs={6}
              >
                <div
                  className="col-6"
                >
                  <div
                    className="box"
                  >
                    <Health
                      displayValue={true}
                      value={84}
                    >
                      <div
                        className="container"
                        style={
                          Object {
                            "padding": 0,
                          }
                        }
                      >
                        <div
                          className="health-value-display"
                        >
                          84
                        </div>
                        <div
                          className="full-bar"
                        >
                          <div
                            className="solid-bar"
                            data-testid="solid-bar"
                            style={
                              Object {
                                "backgroundColor": "#5cb85c",
                                "width": "calc(100% * 0.84)",
                              }
                            }
                          />
                        </div>
                      </div>
                    </Health>
                  </div>
                  <h2>
                    Stats
                  </h2>
                  IP Address: 
                  127.0.0.31
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </Machine>
  </Connect(Machine)>
</Provider>
`;
